[v2.2] Port #22508 and prereqs (ContainerRuntime: Process incoming batches op-by-op instead of waiting for the whole batch) (#22654)
microsoft/FluidFramework: .changeset/metal-hornets-travel.md

---

> "@fluidframework/container-runtime": minor

Author: @tylerbutler should this say `patch` now?

Reviewer: It won't be used regardless, so just leave it as is to make the cherry-pick clean. (We don't use changesets with patches. Maybe we should. Separate discussion.

Author: The cherry-pick is way not clean because of all the churn here.  But I can revert it when I take Jillian's suggestions

Reviewer: It's great to have the changeset, even if unused in this branch.

---
---
"section": "fix"
---

> Restored old op processing behavior around batched ops to avoid potential regression

Reviewer 2: ```suggestion
Restore old op processing behavior around batched ops to avoid potential regression
```

Author: As discussed - Won't Fix (to match 2.3 release notes)

> There's a theoretical risk of indeterminate behavior due to a recent change to how batches of ops are processed.

Reviewer 2: ```suggestion
Due to a recent change to how batches of ops are processed, there's a potential risk of indeterminate sequence numbers between layers.
```
My suggestion may not have the technical details 100% correct, but I think similar wording would be more clear.

Author: As discussed - Won't Fix (to match 2.3 release notes)

This fix reverses that change.

> Pull Request #21785 updated the ContainerRuntime to hold onto the messages in an incoming batch until they've all arrived, and only then process the set of messages.

Reviewer 2: ```suggestion
Pull Request #21785 updated the ContainerRuntime to wait to process a set of messages until all messages from an incoming batch have arrived.
```

Author: As discussed - Won't Fix (to match 2.3 release notes)
